packages/x-charts/src/context/HighlightedProvider/HighlightedProvider.tsx

```diff
@@ -0,0 +1,135 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import useControlled from '@mui/utils/useControlled';
+import {
+  HighlightItemData,
+  HighlightedContext,
+  HighlightScope,
+  HighlightedState,
+  DeprecatedHighlightScope,
+} from './HighlightedContext';
+import { highlightedReducer } from './highlightedReducer';
+import { useSeries } from '../../hooks/useSeries';
+import { ChartSeriesType } from '../../models/seriesType/config';
+import { SeriesId } from '../../models/seriesType/common';
+
+export type HighlightedProviderProps = {
+  children: React.ReactNode;
+  /**
+   * The item currently highlighted. Turns highlighting into a controlled prop.
+   */
+  highlightedItem?: HighlightItemData;
+  /**
+   * The callback fired when the highlighted item changes.
+   *
+   * @param {HighlightItemData} highlightedItem  The newly highlighted item.
+   */
+  onHighlightChange?: (highlightedItem: HighlightItemData) => void;
+};
+
+const mergeDeprecatedOptions = (
+  options?: Partial<HighlightScope> | Partial<DeprecatedHighlightScope>,
+): HighlightScope => {
+  // @ts-expect-error deprecated behavior.
+  const { highlighted, faded, ...rest } = options ?? {};
+  return {
+    highlight: highlighted === 'series' ? 'same-series' : highlighted,
+    fade: faded === 'series' ? 'same-series' : faded,
+    ...rest,
+  };
+};
+
+function HighlightedProvider({
+  children,
+  highlightedItem: highlightedItemProps,
+  onHighlightChange,
+}: HighlightedProviderProps) {
+  const [highlightedItem, setHighlightedItem] = useControlled({
+    controlled: highlightedItemProps,
+    default: null,
+    name: 'HighlightedProvider',
+    state: 'highlightedItem',
+  });
+
+  const [state, dispatch] = React.useReducer(highlightedReducer, {
+    options: undefined,
+    highlightedItem,
+    isFaded: () => false,
+    isHighlighted: () => false,
+  });
+
+  const series = useSeries();
+  const seriesById = React.useMemo(() => {
+    const map: Map<SeriesId, Partial<HighlightScope> | undefined> = new Map();
+
+    Object.keys(series).forEach((seriesType) => {
+      const seriesData = series[seriesType as ChartSeriesType];
+      Object.keys(seriesData?.series ?? {}).forEach((seriesId) => {
+        const seriesItem = seriesData?.series[seriesId];
+        map.set(seriesId, mergeDeprecatedOptions(seriesItem?.highlightScope));
+      });
+    });
+    return map;
+  }, [series]);
+
+  const highlightScope =
+    highlightedItem && highlightedItem.seriesId
+      ? seriesById.get(highlightedItem.seriesId) ?? null
+      : null;
+
+  React.useEffect(() => {
+    dispatch(
+      highlightedItem
+        ? { type: 'set-highlighted', itemData: highlightedItem }
+        : { type: 'clear-highlighted' },
+    );
+  }, [highlightedItem]);
+
+  React.useEffect(() => {
+    dispatch(
+      highlightScope ? { type: 'set-options', options: highlightScope } : { type: 'clear-options' },
+    );
+  }, [highlightScope]);
```

Comment:
I've simplified it up. It was using a reducer because it made a bit more sense in the first iteration, so I just added the `useControlled` on top of it. But you are right that it added a lot of unneeded complexity 👍 

I've left the `"getHighlightedScope"` as is because if we depend on `highlightedItem`, it will run the more costly check every time it changes, which is not necessary, since series don't change much and `seriesById.get(highlightedItem.seriesId)` is a getter with constant time.